Move Clock handling code out of event handlers

frank-weinberg opened the pull request:
Visible changes:
- Before the start of a period the scoreboard will now display Jam 0 instead of Jam 1. This makes the behaviour consistent with any other time where no Jam is on and the number of the previous Jam is displayed.
- Proper fix to #48

Internal changes:
- The code from the button handling functions and event listeners has
been factored into a private set of methods that clearly spell out what each of
them does and run serialized. Only event listeners that trigger
independent of any button presses have remained and those have been
reduced to a single method call.
- With the exception of the period clock all clocks are reset and the
number increased as they are started. A convenience method has been
added to ClockModel to do this in one call. (Period clock is special as
we reset it when the intermission ends, but only start it at the first
Jam.)
- ScoreBoardModel.stopJam() has been renamed to stopJamTO() to better reflect
its current function.
- ScoreBoardModel.timeout() with parameter(s) has been renamed setTimeoutType() and the code that handles the clocks has been separated from the code that handles ownership as far as currently possible. (Since the "Timeout" button currently doubles to mark a timeout as OTO, some ownership code currently has to reside in the clock handling function. Fixing this will require a UI change.)

Closes #48

frank-weinberg committed:
Move Clock handling code out of event handlers

Visible changes:
- Before the start of a period the scoreboard will now display Jam 0 instead of Jam 1. This makes the behaviour consistent with any other time where no Jam is on and the number of the previous Jam is displayed.
- Proper fix to #48

Internal changes:
- The code from the button handling functions and event listeners has
been factored into a private set of methods that clearly spell out what each of
them does and run serialized. Only event listeners that trigger
independent of any button presses have remained and those have been
reduced to a single method call.
- With the exception of the period clock all clocks are reset and the
number increased as they are started. A convenience method has been
added to ClockModel to do this in one call. (Period clock is special as
we reset it when the intermission ends, but only start it at the first
Jam.)
- ScoreBoardModel.stopJam() has been renamed to stopJamTO() to better reflect
its current function.
- ScoreBoardModel.timeout() with parameter(s) has been renamed setTimeoutType() and the code that handles the clocks has been separated from the code that handles ownership as far as currently possible. (Since the "Timeout" button currently doubles to mark a timeout as OTO, some ownership code currently has to reside in the clock handling function. Fixing this will require a UI change.)

frank-weinberg committed:
add changes missed in previous commit

brian-brazil commented on src/com/carolinarollergirls/scoreboard/defaults/DefaultScoreBoardModel.java:
```diff
@@ -489,6 +573,7 @@ protected void createTeamModel(String id) {
 	protected Object timeoutOwnerLock = new Object();
 	protected boolean officialReview;
 	protected Object officialReviewLock = new Object();
+	protected boolean restartPcAfterTimeout;
```
Does this need to be snapshotted?

frank-weinberg replied:
No. Undoing a period preparation (if possible) or a Jam end will always bring us in a state where it will be set again before we can get into a timeout. Changing the timout type is not snapshotted. And no other action sets this variable.
It will however not do any damage to include it and that may save people time wondering if it should be.

brian-brazil replied:
That's what I suspected, mightn't hurt to do it for consistency.

brian-brazil commented on src/com/carolinarollergirls/scoreboard/defaults/DefaultScoreBoardModel.java:
```diff
@@ -185,144 +179,234 @@ public void startJam() {
+		if (tc.isRunning()) {
+			//TODO: Make Official Timeout its own button that calls setTimeoutType()
+			if (getTimeoutOwner()=="") {
 				setTimeoutOwner("O");
```
This convention should be documented somewhere.

frank-weinberg replied:
It's descriebd in the Wiki since 3.9.3.

brian-brazil replied:
Can you point me to where the time out owner can have the value "O" on the wiki? This is an implementation detail I'd expect to be mentioned in a comment, as otherwise I'd expect it to be "1" or "2".

frank-weinberg replied:
Oh, you mean the fact that "O" is OTO. That is indeed not on the Wiki AFAIK.

brian-brazil replied:
It's not in the WS (that comes out as bools), so it probably shouldn't be on the wiki as it's internals only of interest to a few of us. A quick comment somewhere, or appropriately named consts, would do the trick I think.

frank-weinberg replied:
It's in the WS in ScoreBoard.TimeoutOwner and is used in the WS display.

brian-brazil replied:
Ah, I missed that. That's a todo for me then.

brian-brazil commented on src/com/carolinarollergirls/scoreboard/defaults/DefaultScoreBoardModel.java:
```diff
@@ -185,144 +179,234 @@ public void startJam() {
-			ClockModel pc = getClockModel(Clock.ID_PERIOD);
-			ClockModel ic = getClockModel(Clock.ID_INTERMISSION);
-			
+	public void setTimeoutType(String owner, boolean review) {
```
This name implies that this method is setting the type of a timeout that has already started, not actually starting a timeout.

frank-weinberg replied:
Which is what it is used for 99% of the time and what is decribed on the Wiki. That the function also starts a timeout if none is running is a convenience feature.

brian-brazil replied:
I don't understand how a function you have just created can be documented on the wiki. This name is still a bit confusing.

frank-weinberg replied:
This function is triggered when you press one of the Team Timeout or Official Review Buttons on the operator tab. And their function is documented.

brian-brazil replied:
Those buttons are documented, this function is not. My point is that this function name on its own is confusing. I'd expect a function called setTimeoutType to have no effect if no timeout was already in progress as various other functions don't do anything if the scoreboard isn't in the right state.

Please choose a better name, like startTimeoutType.

brian-brazil commented on src/com/carolinarollergirls/scoreboard/defaults/DefaultScoreBoardModel.java:
```diff
@@ -157,17 +153,15 @@ public void startOvertime() {
 			setInOvertime(true);
-			tc.stop();
-			ic.stop();
+			if (tc.isRunning()) {
```
_endTimeout already checks for this, which you rely on elsewhere

brian-brazil commented on src/com/carolinarollergirls/scoreboard/defaults/DefaultScoreBoardModel.java:
```diff
@@ -103,34 +103,31 @@ public void reset() {
 
 	public boolean isInOvertime() { return inOvertime; }
 	public void setInOvertime(boolean o) {
+		if (o == inOvertime) { return; }
```
Do we need to worry about the lc.setMaximumTime getting undone if there's an undo?

frank-weinberg replied:
This is done by setInOvertime(), which restoreSnapshot calls.

brian-brazil commented on src/com/carolinarollergirls/scoreboard/defaults/DefaultScoreBoardModel.java:
```diff
@@ -185,144 +179,234 @@ public void startJam() {
+		}
+		requestBatchEnd();
+	}
+	private void _checkAutostart() {
```
_possiblyAutostartJam would be clearer and consistent with _possiblyEndPeriod. "Check" implies that this function is stateless.

frank-weinberg replied:
The "Jam"-suffix however is too specific, as it might also start a timeout.

brian-brazil commented on src/com/carolinarollergirls/scoreboard/defaults/DefaultScoreBoardModel.java:
```diff
@@ -185,144 +179,234 @@ public void startJam() {
+		ClockModel pc = getClockModel(Clock.ID_PERIOD);
+		ClockModel tc = getClockModel(Clock.ID_TIMEOUT);
+
+		requestBatchStart();
```
You don't close this batch, and the early return will mess this up too. This would break WS at the least.

brian-brazil commented on src/com/carolinarollergirls/scoreboard/defaults/DefaultScoreBoardModel.java:
```diff
@@ -185,144 +179,234 @@ public void startJam() {
-			ClockModel lc = getClockModel(Clock.ID_LINEUP);
+			createSnapshot(ACTION_TIMEOUT);
 
 			requestBatchStart();
```
You use a batch here for a single call that also starts a batch, but not in stopJamTO which does similar.

brian-brazil commented on src/com/carolinarollergirls/scoreboard/defaults/DefaultScoreBoardModel.java:
```diff
@@ -185,144 +179,234 @@ public void startJam() {
-			lc.resetTime();
-			lc.start();
-			requestBatchEnd();
+		if (pc.getTimeRemaining() < 30000) {
```
Should we use the normal lineup time here rather than hardcoding?

frank-weinberg replied:
In the WFTDA rules it is hardcoded to 30s, not duration of lineup.
And the proper way would be to make it configurable in the ruleset. I'll add a TODO note for when I add the other items I have on my list for the rulesets.

brian-brazil replied:
The duration of a lineup is hardcoded to 30s too, so if it's different we're already off the beaten track. Not a problem in practice.

frank-weinberg replied:
The duration of a lineup can be set with the "Clock.Lineup.Time" rule.

brian-brazil reviewed (commented):
This is quite a nice improvement, much clearer what's going on.

How does the "Timeout before period end" function work with all this?

frank-weinberg committed:
review comments

frank-weinberg commented:
That CircleCI failure is interesting. It indicates that the clock is running despite being set to paused in the setup routine.

frank-weinberg commented:
@brian-brazil As far as I am aware, I adressed all your comments. Did I miss anything, you are waiting for or did you just not get around to review again?

brian-brazil commented:
I didn't see that you'd pushed a commit, GitHub doesn't send notifications about that. I'll get to this in the next day.

brian-brazil commented on src/com/carolinarollergirls/scoreboard/defaults/DefaultClockModel.java:
```diff
@@ -313,6 +313,12 @@ public void stop() {
 	}
+	
+	public void startNext() {
+		changeNumber(1);
```
This should probably have a batch around it

brian-brazil reviewed (commented):
Looks like #135 does in fact conflict with this, so I'll wait for this to go in first.

Did you have any other PRs in the works, or could I get the style change in after this?

frank-weinberg committed:
more review comments

Also disable clock sync in unit tests for DefaultScoreBoardModel as it
can cause unwanted clock changes.

frank-weinberg commented:
I have addressed the remaining comments and also found and fixed the source for the failure on the unit tests that looked like clocks were running: When the ClockTimerTask ran for more than 200ms before being paused, clock sync caused the clocks to be modified when started.

brian-brazil commented:
Hmm, I wonder if we might miss stuff long term if the tests aren't using clock sync - but we can deal with that as is happens.
